### PR TITLE
Implement array shiftparsing

### PR DESCRIPTION
**Pull request by alissapajer**
[ch2274]

Depends on #3847 and #3844.

Enables shiftparsing for queries like
```sql
select first.second[_].testField as testField, first.second[_:] as key from `deep-giraffe-array.data`
```

### CODE REVIEW

File: mimir/src/test/scala/quasar/mimir/evaluate/ShiftingSpec.scala [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    import Shifting.{drillToTarget, shiftRValue, ShiftInfo}
+
+    "return entire array when provided path is empty and shift type is array" >> {
+      val target =
+        RArray(List(CString("target"), CDouble(1.2)))
+
+      drillToTarget(target, List(), ShiftType.Array) must equal(Some(target))
+    }
+
+    "return entire object when provided path is empty and shift type is object" >> {
```
Comment (alissapajer) [inline mimir/src/test/scala/quasar/mimir/evaluate/ShiftingSpec.scala]: Added this test which is testing https://github.com/slamdata/quasar/pull/3844. The merge conflicts were going to take too much time to deal with.

File: mimir/src/test/scala/quasar/mimir/evaluate/ShiftingSpec.scala [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+      val shiftPath = ShiftPath(List("three", "three-nested", "three-nested-nested"))
+      val shiftKey = ShiftKey("shiftkey")
+
+      val shiftInfoIncludeId = ShiftInfo(shiftPath, IncludeId, ShiftType.Map, shiftKey)
+      val shiftInfoExcludeId = ShiftInfo(shiftPath, ExcludeId, ShiftType.Map, shiftKey)
+      val shiftInfoIdOnly = ShiftInfo(shiftPath, IdOnly, ShiftType.Map, shiftKey)
+
+      drillToTarget(rvalue, shiftPath.path, ShiftType.Map) must equal(Some(target))
+
+      shiftRValue(rvalue, shiftInfoIncludeId) must equal(
```
Comment (alissapajer) [inline mimir/src/test/scala/quasar/mimir/evaluate/ShiftingSpec.scala]: Also adjusted this test to return multiple results.

File: mimir/src/main/scala/quasar/mimir/evaluate/Shifting.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+object Shifting {
+
+  final case class ShiftInfo(
+    shiftPath: ShiftPath,
+    idStatus: IdStatus,
+    shiftType: ShiftType,
+    shiftKey: ShiftKey)
+
+  /* Returns the `RObject` at the provided path, returning `None` when
+   * the path points to a non-object or the path does not exist.
```
Comment (alissapajer) [inline mimir/src/main/scala/quasar/mimir/evaluate/Shifting.scala]: I'll update this comment to include `RArray`.

File: mimir/src/main/scala/quasar/mimir/evaluate/Shifting.scala [hunk 4]
```diff
@@ -72,6 +79,20 @@ object Shifting {
             case (acc, (_, v)) => RObject((shiftKey, v)) :: acc
           }
       }
+
+      case Some(RArray(elems)) => shiftInfo.idStatus match {
+        case IdOnly =>
+          0.until(elems.length).toList.map(idx => RObject((shiftKey, CLong(idx))))
```
[the remark this answers is not in the document]
Comment (alissapajer) [reply mimir/src/main/scala/quasar/mimir/evaluate/Shifting.scala]: Hm, I can't get this to compile. 
```
Cannot construct a collection of type List[quasar.precog.common.RValue] with elements of type quasar.precog.common.RValue based on a collection of type scala.collection.immutable.IndexedSeq[Int].
```

[the remark this answers is not in the document]
Comment (wemrysi) [reply mimir/src/main/scala/quasar/mimir/evaluate/Shifting.scala]: Oh, that's unfortunate =/

**Comment by alissapajer** [comment]
@wemrysi updated the scaladoc and merged in #3844 . Just waiting on the build.

**Pull request merged**